File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        auto lo{0};
        auto hi{static_cast<int>(heights.size() - 1)};
        auto ret{0};
        while (lo < hi) {
            ret = max(ret, min(heights[lo], heights[hi]) * (hi - lo));
            if (heights[lo] < heights[hi]) {
                lo++;
            } else {
                hi--;
            }
        }
        return ret;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-2.py
class Solution:
    def trap(self, height: List[int]) -> int:
        maxL = [0] * len(height)
        maxR = [0] * len(height)

        tempMax = 0
        for i, h in enumerate(height):
            if i != 0:
                maxL[i] = tempMax
            tempMax = max(tempMax, h)
        
        tempMax = 0
        for i, h in enumerate(reversed(height)):
            if i != 0:
                maxR[len(height) - i - 1] = tempMax
            tempMax = max(tempMax, h)
        
        print(maxL)
        print(maxR)
        tot = 0
        for i in range(len(height)):
            if height[i] < min(maxL[i], maxR[i]):
                tot += min(maxL[i], maxR[i]) - height[i]
        return tot

